xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/store/XWikiVersioningStoreInterface.java

```diff
@@ -47,6 +55,74 @@ void saveXWikiDocArchive(XWikiDocumentArchive archivedoc, boolean bTransaction,
 
     Version[] getXWikiDocVersions(XWikiDocument doc, XWikiContext context) throws XWikiException;
 
+    private Collection<String> getXWikiDocStringVersions(XWikiDocument doc, RevisionCriteria criteria,
+        XWikiContext context) throws XWikiException
+    {
+        List<String> results = new ArrayList<>();
+
+        Version[] revisions = getXWikiDocVersions(doc, context);
+        XWikiDocumentArchive archive = getXWikiDocumentArchive(doc, context);
+
+        XWikiRCSNodeInfo nextNodeinfo = null;
+        XWikiRCSNodeInfo nodeinfo;
+        for (Version revision : revisions) {
+            nodeinfo = nextNodeinfo;
+            nextNodeinfo = archive.getNode(revision);
+
+            if (nodeinfo != null) {
+                // Minor/Major version matching
+                if (criteria.getIncludeMinorVersions() || nextNodeinfo == null || !nextNodeinfo.isMinorEdit()) {
+                    // Author matching
+                    if (criteria.getAuthor().isEmpty() || criteria.getAuthor().equals(nodeinfo.getAuthor())) {
+                        // Date range matching
+                        Date versionDate = nodeinfo.getDate();
+                        if (versionDate.after(criteria.getMinDate()) && versionDate.before(criteria.getMaxDate())) {
+                            results.add(nodeinfo.getVersion().toString());
+                        }
+                    }
+                }
+            }
+        }
+
+        return criteria.getRange().subList(results);
+    }
+
+    /**
+     * Gets versions from a given document matching criteria like author, minimum creation date, etc.
+     *
+     * @param doc the document
+     * @param criteria criteria used to match versions
+     * @param context the XWiki context
+     * @return the matching versions
+     * @since 15.10.7
+     * @since 16.2.0RC1
+     */
+    @Unstable
+    default Collection<Version> getXWikiDocVersions(XWikiDocument doc, RevisionCriteria criteria, XWikiContext context)
+        throws XWikiException
+    {
+        return getXWikiDocStringVersions(doc, criteria, context).stream()
```

Comment:
Honestly not sure I'm a big fan of having a whole implemented method in the interface just for a fallback that will be very unlikely used: I mean we provide a default just in case someone would have maybe re-implemented this interface. Which might be very very rare. So I would only return an empty collection here. And same below I'd only return 0. 
Providing an implem here means that we'll need to maintain it on the long run, while we don't really want to: it should be maintained in the actual implem, not in the interface IMO.

Reply:
Check the warning in https://dev.xwiki.org/xwiki/bin/view/Community/DevelopmentPractices#HBackwardCompatibility it says:
> Note that the default method should not throw an exception in general (like throw new UnsupportedOperationException("Not supported");), since that would mean that the code calling the default method will fail. As such, it cannot be considered as backward compatibility.

Reply:
Indeed. I'm still unsure on what to do then, I understand your point but at the same time I'm not a fan of returning empty values.

Reply:
I prefer the current version (real retro compatibility). The only alternative if to not provide any default implementation so that this breakage is documented in the release note (but I'm really not a fan since we have a retro compatible solution).

Reply:
:+1: for the real retro compatibility. What we should make sure is that it is covered by a unit test (I haven't checked if it is) to ensure that it is actually working as this is code that will never be executed in any real use and it would be quite pointless if it should turn out that it doesn't actually work as it was never tested.

Reply:
There are still no tests, as I wanted feedback on the implementation first. I'll add them tomorrow.